File: atcoder/past201912_e.py
# https://atcoder.jp/contests/past201912/tasks/past201912_e

import copy
N ,Q = map(int, input().split())
f = [[0]* N for _ in range(N)]

for q in range(Q):
    S = list(map(int, input().split()))
    if S[0] == 1:
        f[S[1]-1][S[2]-1] = 1
    elif S[0] == 2:
        for i in range(N):
            if f[i][S[1]-1] == 1:
                f[S[1]-1][i] = 1
    else:
        # print(f)
        tmp = copy.copy(f[S[1]-1])
        for j in range(N):
            if tmp[j] == 1:
                # print(f[S[1]-1][j], S[1]-1, j, f[0])
                for k in range(N):
                    if f[j][k] == 1 and S[1]-1 != k:
                        f[S[1]-1][k] = 1
            else:
                continue
        # print(f)

# print(f)
for i in range(N):
    ans = ''
    for j in range(N):
        if f[i][j]:
            ans += 'Y'
        else:
            ans += 'N'
    print(ans)

File: atcoder/past201912_e.cpp
// https://atcoder.jp/contests/past201912-open/tasks/past201912_e
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
#include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define all(v) (v).begin(),(v).end()
#define chmin(x,y) x = min(x,y)
#define chmax(x,y) x = max(x,y)
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;
ll gcd(ll a, ll b) { return b?gcd(b,a%b):a;}
ll lcm(ll a, ll b) { return a/gcd(a,b)*b;}

int main() {
	int n, q;
	cin >> n >> q;
    vector<vector<int>> f(n, vector<int>(n, 0));
    int type, from, to;
    rep(i,q) {
        cin >> type;
        if (type==1) {
            cin >> from >> to;
            from--; to--;
            f[from][to] = 1;
        } else if (type==2) {
            cin >> from;
            from--;
            rep(j,n) if(f[j][from] && from != j) f[from][j] = 1;
        } else {
            cin >> from;
            from--;
            vector<int> v;
            rep(j,n) if(f[from][j]) v.push_back(j);
            for(auto j : v)  rep(k,n) if(f[j][k] && from != k) f[from][k] = 1;
        }
    }
    rep(i,n) {
        rep(j,n) {
            if (f[i][j]) cout << 'Y';
            else         cout << 'N';
        }
        cout << endl;
    }
	return 0;
}
